src/Nethermind/Nethermind.Serialization.Rlp/BlockDecoder.cs

```diff
@@ -55,16 +55,29 @@ public class BlockDecoder : IRlpValueDecoder<Block>, IRlpStreamDecoder<Block>
 
             if (rlpStream.Position != blockCheck)
             {
-                int withdrawalsLength = rlpStream.ReadSequenceLength();
-                int withdrawalsCheck = rlpStream.Position + withdrawalsLength;
-                withdrawals = new();
-
-                while (rlpStream.Position < withdrawalsCheck)
+                bool lengthWasRead = true;
+                try
+                {
+                    rlpStream.PeekNextRlpLength();
+                }
+                catch
                 {
-                    withdrawals.Add(Rlp.Decode<Withdrawal>(rlpStream));
+                    lengthWasRead = false;
                 }
 
-                rlpStream.Check(withdrawalsCheck);
+                if (lengthWasRead)
```

Comment:
We do not know if withdrawals already enabled or not, that is why we have these ugly check (most likely). And it would be better to make a method TryPeekNextRlpLength instead of handling costly exceptions.

Reply:
TL:DR - it is fine, sorry for the confusion

In case of disabled withdrawals and no additional bytes, we will have `false` in `if (rlpStream.Position != blockCheck)` and we will not be inside this part of code. If we are here it means that there is something in block RLP after uncles. It can be:
- withdrawals
- or some additional bytes.

If we are not able to read length, it means that it's not valid-formatted withdrawals, but only some additional bytes. So at this point we need to check if additional bytes are allowed.
And we are doing it just after `if (rlpStream.Position != blockCheck)` so it's perfectly fine! We will go to `rlpStream.Check(blockCheck)` and throw anyway

Reply:
@flcl42 yeah that's checks are super ugly. Why can't we assume that withdrawals are enabled before 4844?